test/unit/math/mix/fun/bessel_first_kind_test.cpp

```diff
@@ -13,3 +13,19 @@ TEST(mathMixScalFun, besselFirstKind) {
   stan::test::expect_ad(f(1), std::numeric_limits<double>::quiet_NaN());
   stan::test::expect_ad(f(2), 2.79);
 }
+
+TEST(mathMixScalFun, besselFirstKind_vec) {
+  auto f = [](const auto& x1, const auto& x2) {
+    using stan::math::bessel_first_kind;
+    return bessel_first_kind(x1, x2);
+  };
+
+  std::vector<int> std_in1{3, 1};
+  Eigen::VectorXd in2(2);
+  in2 << 0.5, 3.4;
+  stan::test::expect_ad_vectorized_binary(f, std_in1, in2);
```

Comment:
Does this also test if in2 is an Eigen row vector or a std::vector type?

A second question is, what signatures are we exposing at the Stan level?

Is it just:

```
vector bessel_first_kind(int [] i, real a);
vector bessel_first_kind(int [] i, vector a);
vector bessel_first_kind(int i, vector a);
```

Or does it also have row vector types?
```
vector bessel_first_kind(int i, row_vector a);
```

Or something.

Reply:
> Does this also test if in2 is an Eigen row vector or a std::vector type?

Not at the moment, should I add it?

> A second question is, what signatures are we exposing at the Stan level?

I was supposed to update the top-level description, sorry. This will expose the following signatures:
```
real bessel_first_kind(int v, real x)
real[] bessel_first_kind(int[] v, real[] x)
real[] bessel_first_kind(int v, real[] x)
real[] bessel_first_kind(int[] v, real x)
vector bessel_first_kind(int[] v, vector x)
vector bessel_first_kind(int v, vector x)
row_vector bessel_first_kind(int[] v, row_vector x)
row_vector bessel_first_kind(int v, row_vector x)
matrix bessel_first_kind(int[,] v, matrix x)
matrix bessel_first_kind(int v, matrix x)
```

These can also be arbitrarily nested (depending the level of signature we want to expose), as long as both arguments have the same level of nesting or one is a scalar:
```
real[,] bessel_first_kind(int[,] v, real[,] x)
real[,] bessel_first_kind(int v, real[,] x)
real[,] bessel_first_kind(int[,] v, real x)
vector[] bessel_first_kind(int[,] v, vector[] x)
vector[] bessel_first_kind(int v, vector[] x)
row_vector[] bessel_first_kind(int[,] v, row_vector[] x)
row_vector[] bessel_first_kind(int v, row_vector[] x)
matrix[] bessel_first_kind(int[,,] v, matrix[] x)
matrix[] bessel_first_kind(int v, matrix[] x)

```

Reply:
Oh okay so it was prohibitively slow to have all the combos in the test framework. That's less than ideal but makes sense.

I feel like we should have at least one test per signature we're exposing at the Stan level. Can you do this with prim for these functions? Like you can make these tests pretty easily. Just define an `Eigen::VectorXd`, an `Eigen::RowVectorXd`, a `std::vector` and fill them with the same values and call them with the arguments in the forward order + reversed and just check the values are always the same as some reference (use `EXPECT_MATRIX_FLOAT_EQ`).

Or I guess that could go in the prim binary testing bit? However is fine. I think we should at least have code in there checking values + checking that these things compile.

I'm a little hesitant about going crazy with the nesting -- only need so many signatures here -- I'd be happy with 1D things. This sounds like an extension of some vectorization that's already in. Presumably just match whatever you did for that.

Reply:
> Yep, my original goal was to have the same testing that was used in the prim vectorisation tests, which test:

Oh wait I just read your comments again. Are these being tested with all these different types in prim already?

Reply:
Yep that's being tested in ```prim```, but I realised I need to expand the testing with ```std::vector<int>``` inputs, so now it's testing:
```
bessel_first_kind(int, real)
bessel_first_kind(int, vector)
bessel_first_kind(int, row_vector)
bessel_first_kind(int, matrix)
bessel_first_kind(int, vector[])
bessel_first_kind(int, row_vector[])
bessel_first_kind(int, matrix[])
bessel_first_kind(int[], real)
bessel_first_kind(int[], vector)
bessel_first_kind(int[], row_vector)
bessel_first_kind(int[,], matrix)
```